src/solver/impls/snes/snes.cxx

```diff
@@ -1291,8 +1291,16 @@ void SNESSolver::updateColoring() {
   // Replace the old coloring with the new one
   MatFDColoringDestroy(&fdcoloring);
   MatFDColoringCreate(Jfd, iscoloring, &fdcoloring);
-  MatFDColoringSetFunction(
-      fdcoloring, reinterpret_cast<PetscErrorCode (*)()>(FormFunctionForColoring), this);
+  MatFDColoringSetFunction(fdcoloring,
+#if PETSC_VERSION_GE(3, 24, 0) \
+    || (PETSC_VERSION_GE(3, 23, 0) && PETSC_VERSION_RELEASE == 0)
+                           reinterpret_cast<PetscErrorCode (*)(void*, Vec, Vec, void*)>(
+                               FormFunctionForColoring),
+#else
+                           reinterpret_cast<PetscErrorCode (*)()>(
+                               FormFunctionForColoring),
+#endif
```

Comment:
I'd really _very_ much prefer it if we could keep preprocessor guards out of expressions. It looks like PETSc 3.23 provides a `MatFDColoringFn` typedef, so instead, we could provide a backport of this in our petsc shim header:

```cpp
// Or whatever the correct version is
#if PETSC_VERSION_LE(3, 23, 0)
using MatFDColoringFn = PetscErrorCode (*)();
#endif
```

and then just use that here (and elsewhere):

```suggestion
                           reinterpret_cast<MatFDColoringFn>(
                               FormFunctionForColoring
                           ),
```